src/hpc/aobjects.c

```diff
@@ -2029,6 +2040,27 @@ static Int InitKernel (
   AssRecFuncs[ T_TLREC ] = AssTLRecord;
   IsRecFuncs[ T_TLREC ] = AlwaysYes;
   UnbRecFuncs[ T_TLREC ] = UnbTLRecord;
+
+  // Ensure that atomic objects cannot be copied or compared
+  // as this is impossible to do thread-safely
+  for (UInt type = FIRST_ATOMIC_TNUM; type <= LAST_ATOMIC_TNUM; type++) {
+      ShallowCopyObjFuncs[type] = AtomicErrorNoShallowCopy;
+      CopyObjFuncs[type] = AtomicErrorNoCopy;
+      // Do not error on CleanObj, just leave it as a no-op
+
+      for (UInt t2 = FIRST_REAL_TNUM; t2 <= LAST_REAL_TNUM; ++t2) {
+          // Allow the common case of comparing against booleans.
+          // This is safe to allow, as it doesn't actually inspect
+          // the atomic objects, and allows comparing to 'fail'.
+          if (t2 != T_BOOL) {
+              EqFuncs[type][t2] = AtomicErrorNoCompare;
+              EqFuncs[t2][type] = AtomicErrorNoCompare;
+              LtFuncs[type][t2] = AtomicErrorNoCompare;
+              LtFuncs[t2][type] = AtomicErrorNoCompare;
```

Comment:
Hmm, I don't understand why it is necessary to be so drastic: the default `EqFuncs` for objects whose TNUM differs is `EqNot`, which is perfectly thread safe. I.e. using `=` to copare an atomic record with, say, an int, a perm or a FFE should be just as safe as comparing them to a bool. Why override them then?

I have no problem with the change to `LtFuncs`, though, as I personally would prefer if most pairs of objects of differing type were not comparable with `<` :-)

Reply:
My main aim was to make sure you can't compare to any of the list types, as there it is confusing if `[1,2,3] = AtomicList([1,2,3])` was false. I'm happy to scale back to just the list types.

Reply:
And obviously also don't compare atomic records with normal records.

Reply:
So the question is (sorry for lots of bitty replies), would you like to just ban atomic lists with other lists, and atomic records with other records, and leave everything else false?

Reply:
I'd indeed disallow comparing atomic lists/records to any other kinds of (atomic or non-atomic) lists or records, as the user may (incorrectly) expect them to do something, so just returning `false` could be misleading.
OTOH, comparing them to any non-list or non-record can simply return `false` as it does right now, as there is no way these objects could be equal.

I guess strictly speaking this leaves the comparison with external objects, as those could be list/record objects, too... if we want to be on the safe said, this should therefore produce an error, too. I don't really care much about those, though; my primary concern is that I think producing an error when comparing an atomic list to an integer is more confusing than helpful. But that's just me, of course, others might disagree.